options/scripting_python.go

```diff
@@ -0,0 +1,100 @@
+package options
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+)
+
+// ScriptingPython defines the configuration of a python environment.
+type ScriptingPython struct {
+	VirtualEnvPath        string            `bson:"virtual_env_path" json:"virtual_env_path" yaml:"virtual_env_path"`
+	RequirementsFilePath  string            `bson:"requirements_path" json:"requirements_path" yaml:"requirements_path"`
+	HostPythonInterpreter string            `bson:"host_python" json:"host_python" yaml:"host_python"`
+	Packages              []string          `bson:"packages" json:"packages" yaml:"packages"`
+	LegacyPython          bool              `bson:"legacy_python" json:"legacy_python" yaml:"legacy_python"`
+	CachedDuration        time.Duration     `bson:"cache_duration" json:"cache_duration" yaml:"cache_duration"`
+	Environment           map[string]string `bson:"env" json:"env" yaml:"env"`
+	Output                Output            `bson:"output" json:"output" yaml:"output"`
+
+	requirementsMTime time.Time
+	cachedAt          time.Time
+	requrementsHash   string
+	cachedHash        string
+}
+
+// NewPythonScriptingEnvironmnet generates a ScriptingEnvironment
+// taking the arguments given for later use. Use this function for
+// simple cases when you do not need or want to set as many aspects of
+// the environment configuration.
+func NewPythonScriptingEnvironmnet(path, reqtxt string, packages ...string) ScriptingEnvironment {
+	return &ScriptingPython{
+		CachedDuration:        time.Hour,
+		HostPythonInterpreter: "python3",
+		Packages:              packages,
+		VirtualEnvPath:        path,
+		RequirementsFilePath:  reqtxt,
+	}
+}
+
+func (opts *ScriptingPython) Interpreter() string {
+	return filepath.Join(opts.VirtualEnvPath, "bin", "python")
+}
+
+func (opts *ScriptingPython) Type() string { return "python" }
+
+func (opts *ScriptingPython) Validate() error {
+	if opts.CachedDuration == 0 {
+		opts.CachedDuration = 10 * time.Minute
+	}
+
+	if opts.VirtualEnvPath == "" {
+		opts.VirtualEnvPath = filepath.Join("venv", uuid.Must(uuid.NewV4()).String())
+	}
+
+	if opts.HostPythonInterpreter == "" {
+		opts.HostPythonInterpreter = "python3"
+	}
+
+	return nil
+}
+
+func (opts *ScriptingPython) ID() string {
```

Comment:
This is potentially racy. Someone else could call ID and write to the cache. Is that a concern?

Reply:
yeahhhhhhhhh, I'm not super worried about it. 

I think the hashes will be stable. I'm more worried about someone having an options struct, creating it, losing track of the hash, and then changing it, and creating a different Environment, which I don't think will be a problem in reasonable situations.